getting-started/bookshelf/main.go

```diff
@@ -270,32 +275,54 @@ func (b *Bookshelf) deleteHandler(w http.ResponseWriter, r *http.Request) *appEr
 	ctx := r.Context()
 	id := mux.Vars(r)["id"]
 	if err := b.DB.DeleteBook(ctx, id); err != nil {
-		return appErrorf(err, "DeleteBook: %v", err)
+		return b.appErrorf(r, err, "DeleteBook: %v", err)
 	}
 	http.Redirect(w, r, "/books", http.StatusFound)
 	return nil
 }
 
+func (b *Bookshelf) sendLog(w http.ResponseWriter, r *http.Request) *appError {
+	b.logger.Println("Hey, you triggered a custom log entry. Good job!")
+
+	fmt.Fprintln(w, "Log sent! Check the logging section of the Cloud Console.")
+
+	return nil
+}
+
+func (b *Bookshelf) sendError(w http.ResponseWriter, r *http.Request) *appError {
+	fmt.Fprintf(w, "Logging an error. Check Error Reporting.")
+	err := errors.New("uh oh! an error occurred")
+	return b.appErrorf(r, err, "%v", err)
+}
+
 // https://blog.golang.org/error-handling-and-go
 type appHandler func(http.ResponseWriter, *http.Request) *appError
 
 type appError struct {
-	Error   error
-	Message string
-	Code    int
+	err     error
+	message string
+	code    int
+	req     *http.Request
+	b       *Bookshelf
 }
 
 func (fn appHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	if e := fn(w, r); e != nil { // e is *appError, not os.Error.
-		log.Printf("Handler error: status code: %d, message: %s, underlying err: %#v", e.Code, e.Message, e.Error)
-		http.Error(w, e.Message, e.Code)
+		e.b.errorClient.Report(errorreporting.Entry{
+			Error: e.err,
+			Req:   r,
+		})
+		e.b.errLogger.Printf("Handler error: status code: %d, message: %s, underlying err: %#v", e.code, e.message, e.err)
+		http.Error(w, e.message, e.code)
 	}
 }
 
-func appErrorf(err error, format string, v ...interface{}) *appError {
+func (b *Bookshelf) appErrorf(r *http.Request, err error, format string, v ...interface{}) *appError {
```

Comment:
Passing around the bookshelf object seems reasonable and useful, but it looks to me like maybe it should be embedded inside a context object? That seems to be the kind of role you have in mind for it.

Reply:
I need access to `errorClient` and `errLogger` in `ServeHTTP`. Neither of those is request-scoped, so I don't think it's a good fit to embed in the context - could be wrong. The `errLogger` is kind of request scoped in that we want to pass it the request info, but I'm not sure if that works here.